### PR TITLE
explainer/spec: Sync inputs must await each item

### PR DESCRIPTION
**Pull request by js-choi**
Fixes #9. See also #7.

CC @zloirock.

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -48,14 +48,17 @@ <h1><ins>Array.fromAsync ( _asyncItems_ [ , _mapfn_ [ , _thisArg_ ] ] )</ins></h
             1. Else,
               1. If IsCallable(_mapfn_) is *false*, throw a *TypeError* exception.
               1. Let _mapping_ be *true*.
-            1. Let _usingAsyncOrSyncIterator_ be ? GetMethod(_asyncItems_, @@asyncIterator).
-            1. If _usingAsyncOrSyncIterator_ is *undefined*, let _usingAsyncOrSyncIterator_ be ? GetMethod(_asyncItems_, @@iterator).
+            1. Let _usingIterator_ be ? GetMethod(_asyncItems_, @@asyncIterator).
+            1. If _usingIterator_ is *undefined*, then
+              1. Set _usingIterator_ to ? GetMethod(_asyncItems_, @@iterator).
+              1. If _usingIterator_ is *undefined*, set _usingIterator_ to CreateArrayIterator(_asyncItems_, ~value~).
```
Comment (zloirock) [inline spec.html:54]: It seems missed `GetIterator` steps for async and sync iterators. In the cases of async and sync iterator, `asyncItems` is a **method**, in the case of non-iterable array-like, `asyncItems` is an **iterator**. `CreateAsyncFromSyncIterator` should accept an **iterator record**.

Comment (js-choi) [reply spec.html]: Yeah, I should have waited until I got some sleep to push this. Thanks for pointing out this mistake. I just pushed another commit, which probably should be correct now. (I’m trying to preserve the order in which the arguments’ methods get called, to be the same as with `Array.from`.)

File: spec.html [hunk 2]
```diff
@@ -48,59 +48,109 @@ <h1><ins>Array.fromAsync ( _asyncItems_ [ , _mapfn_ [ , _thisArg_ ] ] )</ins></h
...
-            1. If _usingAsyncOrSyncIterator_ is not *undefined*, then
-              1. If IsConstructor(_C_) is *true*, then
-                1. Let _A_ be ? Construct(_C_).
-              1. Else,
-                1. Let _A_ be ! ArrayCreate(0).
-              1. Let _iteratorRecord_ be ? GetIterator(_asyncItems_, ~async~, _usingAsyncOrSyncIterator_).
+            1. Let _usingAsyncIterator_ be ? GetMethod(_asyncItems_, @@asyncIterator).
+            1. Else,
+              1. Let _usingSyncIterator_ be ? GetMethod(_asyncItems_, @@iterator).
+              1. If _usingSyncIterator_ is *undefined*, set _usingIterator_ to CreateArrayIterator(_asyncItems_, ~value~).
```
Comment (zloirock) [inline spec.html:54]: After that, `usingSyncIterator` could be a function or an iterator. It seems that `CreateArrayIterator` fallback should be moved below.

Comment (js-choi) [reply spec.html]: Fixed in 8b8f766df77f0dd60cc36210619810ea080ba307.

File: spec.html [hunk 3]
```diff
@@ -48,59 +48,109 @@ <h1><ins>Array.fromAsync ( _asyncItems_ [ , _mapfn_ [ , _thisArg_ ] ] )</ins></h
...
+            1. If IsConstructor(_C_) is *true*, then
+              1. Let _A_ be ? Construct(_C_).
+            1. Else,
+              1. Let _A_ be ! ArrayCreate(0).
+            1. If _usingAsyncIterator_ is not *undefined*, then
+              1. Let _iteratorRecord_ be ? GetIterator(_asyncItems_, ~async~, _usingAsyncIterator_).
+            1. Else if _usingSyncIterator_ is not *undefined*_usingAsyncIterator_, then
+              1. Let _iteratorRecord_ be ? CreateAsyncFromSyncIterator(GetIterator(_syncItems_, ~sync~, _usingSyncIterator_)).
+            1. Else,
+              1. Set _iteratorRecord_ to CreateAsyncFromSyncIterator(CreateArrayIterator(_asyncItems_, ~value~)).
```
Comment (zloirock) [inline spec.html:62]: One more minor problem: `CreateAsyncFromSyncIterator` accepts an iterator record, not just an iterator. `GetIterator` returns an iterator record. But `CreateArrayIterator` returns just an iterator.

Comment (zloirock) [reply spec.html:62]: You could use a fallback to `%Array.prototype.values%` + `GetIterator` (like [here](https://github.com/tc39/proposal-array-from-async/commit/64a49214f6a7036e7622ff4b4b9f938e277ae7ae#r56716124)) or create the record manually.

Comment (js-choi) [reply spec.html]: Hopefully e151ba1192b0fdbcf593ca619ba818fb337d6b89 fixes this.

**Review by zloirock — APPROVED** [review]

**Pull request merged**